### PR TITLE
pkg/utensor: add support to RIOT

### PR DESCRIPTION
**Pull request by aabadie**
### Contribution description

This PR adapts the [uTensor library](https://github.com/uTensor/uTensor) to RIOT as a package.

The sample application provided in the tests is just an adaption of the example application given in this [blog post](https://www.hackster.io/news/simple-neural-network-on-mcus-a7cbd3dc108c).
The application is demonstrating how to load a model (model structure + weights) generated from a trained model to determine a digit in an image.
The model is trained using TensorFlow on the MNIST dataset.

The digit image is loaded in the firmware using the recently introduced blob mechanism. There is a helper Python script added to be able to update the digit binary file easily: this script load the test data from the MNIST dataset and, given the index in the list of image, save the image on disk and display its content. This is handy for the user to be able to compare with the value detected by the firmware.

The model files (c++) are automatically generated by the utensor-cli tool and stored in the `models` subdirectory of the application. These C++ files (model description + weights) are loaded in the application as an external module.

The resulting firmware is quite huge, ~180kiB ROM size, and the package requires C++ support, so a few boards (AVR, MSP430, hifive) cannot be used. It should work on MIPS :)

### Testing procedure

The following command should succeed:
```
make BOARD=<board of your choice> -C tests/pkg_utensor flash test
```

### Issues/PRs references

None

### CODE REVIEW

**Comment by kaspar030** [comment]
force-pushing a lot...

**Comment by aabadie** [comment]
> force-pushing a lot...

Indeed, I'm trying to fix the build issues reported by Murdock. Since no one reviewed yet, I think that's ok :)

File: pkg/utensor/patches/0002-fix-mbed-specific-code.patch [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+diff --git a/src/uTensor/util/uTensor_util.hpp b/src/uTensor/util/uTensor_util.hpp
+index 25ebcf3..9fff08b 100644
+--- a/src/uTensor/util/uTensor_util.hpp
++++ b/src/uTensor/util/uTensor_util.hpp
+@@ -107,7 +107,7 @@ void printVector(std::vector<uint32_t> vec);
+     #undef round
+     #undef abs
+ #else
+-    #include "mbed.h"
++    // #include "mbed.h"
```
[the remark this answers is not in the document]
Comment (aabadie) [reply pkg/utensor/patches/0002-fix-mbed-specific-code.patch]: Removed

File: pkg/utensor/patches/0001-fix-integer-format.patch [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+diff --git a/src/uTensor/core/context.cpp b/src/uTensor/core/context.cpp
+index 112308e..e2d04e6 100644
+--- a/src/uTensor/core/context.cpp
++++ b/src/uTensor/core/context.cpp
+@@ -3,7 +3,7 @@
+ S_TENSOR Context::add(Tensor* t, TName _name, uint8_t init_count) {
+   if(t == nullptr) { ERR_EXIT("null pointer tensor"); }
+   if(rTable.find(_name) != rTable.end()) {
+-    ERR_EXIT("tensor with name \"%d\" address already exist in rTable", t->getName().get_value());
++    ERR_EXIT("tensor with name \"%d\" address already exist in rTable", (unsigned)t->getName().get_value());
```
[the remark this answers is not in the document]
Comment (aabadie) [reply pkg/utensor/patches/0001-fix-integer-format.patch]: All fixed

**Comment by aabadie** [comment]
Thanks for your initial review @josephnoir. I addressed all your comments. I also noticed that the utensor-cli (used to generated the model/weights ccp files) has been updated but the version on pypi is now broken so I opened https://github.com/uTensor/utensor_cgen/pull/97 to provide a fix.

**Comment by aabadie** [comment]
Since the test application provides an automatic test script, I ran Murdock with `CI: run tests` enabled and there was only [one unrelated failure](https://ci.riot-os.org/RIOT-OS/RIOT/12805/8a9ded984dbf9d98786b4aeee408247a7af28bcf/output/run_test/tests/pkg_relic/nrf52dk:gnu.txt).

**Pull request merged**